src/main/rx/srxl2.c

```diff
@@ -0,0 +1,610 @@
+
+            if (unitId == 0) {
+                state = SendHandshake;
+                timeoutTimestamp = now + SRXL2_SEND_HANDSHAKE_TIMEOUT_US;
+                fullTimeoutTimestamp = now + SRXL2_LISTEN_FOR_HANDSHAKE_TIMEOUT_US;
+            } else {
+                state = ListenForHandshake;
+                timeoutTimestamp = now + SRXL2_LISTEN_FOR_HANDSHAKE_TIMEOUT_US;
+            }
+        }
+    } break;
+
+    case SendHandshake: {
+        if (now >= timeoutTimestamp) {
+            // @todo set another timeout for 50ms tries
+            // fill write buffer with handshake frame
+            result |= RX_FRAME_PROCESSING_REQUIRED;
+        }
+        // else if (handshake_received) {
+        //  set baud rate accordingly
+        //  state = Running
+        // }
+
+        if (now >= fullTimeoutTimestamp) {
+            serialSetBaudRate(serialPort, SRXL2_PORT_BAUDRATE_DEFAULT);
+            //DEBUG("case SendHandshake: switching to %d baud\r\n", SRXL2_PORT_BAUDRATE_DEFAULT);
+            timeoutTimestamp = now + SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT_US;
+            result = (result & ~RX_FRAME_PENDING) | RX_FRAME_FAILSAFE;
+
+            state = ListenForActivity;
+            lastReceiveTimestamp = 0;
+        }
+    } break;
+
+    case ListenForHandshake: {
+        if (now >= timeoutTimestamp)  {
+            serialSetBaudRate(serialPort, SRXL2_PORT_BAUDRATE_DEFAULT);
+            //DEBUG("case ListenForHandshake: switching to %d baud\r\n", SRXL2_PORT_BAUDRATE_DEFAULT);
+            timeoutTimestamp = now + SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT_US;
+            result = (result & ~RX_FRAME_PENDING) | RX_FRAME_FAILSAFE;
+
+            state = ListenForActivity;
+            lastReceiveTimestamp = 0;
+        }
+    } break;
+
+    case Running: {
+        // frame timed out, reset state
+        if (lastValidPacketTimestamp < now && now - lastValidPacketTimestamp >= SRXL2_FRAME_TIMEOUT_US) {
+            serialSetBaudRate(serialPort, SRXL2_PORT_BAUDRATE_DEFAULT);
+            //DEBUG("case Running: switching to %d baud: %d %d\r\n", SRXL2_PORT_BAUDRATE_DEFAULT, now, lastValidPacketTimestamp);
+            timeoutTimestamp = now + SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT_US;
+            result = (result & ~RX_FRAME_PENDING) | RX_FRAME_FAILSAFE;
+
+            state = ListenForActivity;
+            lastReceiveTimestamp = 0;
+            lastValidPacketTimestamp = 0;
+        }
+    } break;
+    };
+
+    if (writeBufferIdx) {
+        result |= RX_FRAME_PROCESSING_REQUIRED;
+    }
+
+    return result;
+}
+
+static bool srxl2ProcessFrame(const rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    UNUSED(rxRuntimeConfig);
+
+    if (writeBufferIdx == 0) {
+        return true;
+    }
+
+    const uint32_t now = micros();
+
+    if (lastIdleTimestamp > lastReceiveTimestamp) {
+        // time sufficient for at least 2 characters has passed
+        if (now - lastReceiveTimestamp > SRXL2_REPLY_QUIESCENCE) {
```

Comment:
I've done some tests with telemetry getting handled sooner than it currently does, and this part is still required to guarantee the idle period required. With the speed at which telemetry is currently handled however, it doesn't seem to be necessary, but I'd like to keep this in case that changes

Reply:
You can set idle timeout delay to value equal or slightly smaller than telemetry delay. Then it is always possible to transmit just after timeout ISR

Reply:
Could you elaborate more on how to do this? It sounds like you're saying there is a way to configure the idle timeout for the uart peripheral, but I cant find any registers to do so. Based on my interpretation of the STM32 documentation, the idle interrupt delay is fixed to one byte worth of time where the signal line is idle (high for non inverted uart). This means the delay will also vary based on baud rate.

Reply:
I opened F3 documentation I have locally, USARTx_RTOR is available. It is possible cores on different CPUs don't support this (and IMO F1 doesn't support it at all)

Reply:
I was referencing the reference manual for the STM32F405 and I cant find any references to that register. It seems it is only supported on some MCU's then.  With that given, would this existing implementation be acceptable, or do you have another suggestion on how this could be addressed better?

Reply:
Yes, peripheral codes are different. F3 is analog-oriented, so it's cores are usually more advanced.

There isn't much to improve now ... 
just write is as `if (cmpTimeUs(now, lastReceiveTimestamp) > SRXL2_REPLY_QUIESCENCE)` 
or `if (lastReceiveTimestamp && cmpTimeUs(now, lastReceiveTimestamp) > SRXL2_REPLY_QUIESCENCE)` if lastReceiveTimestamp may be zeroed until first packet is received (but first version will fail only if no packet was received for about 30 minutes after power up)

`cmpTimeUs` is just wrapper around unsigned subtraction converted to signed,  but it's obvious that timestamps are compared. Resulting value is signed difference between timestamps, with range of about +-30 minutes (2^31 uS) and correct timestamp wrap handling 
(simple subtraction is not portable; cmp implementation is also not portable, but works on ARM (and most modern processors) and can be easily fixed in single place is necessary)

Reply:
your first  suggestion should work fine. `lastReceiveTimestamp` is set per byte reception so packetes dont have to be valid, and assuming a receiver is connected physically, it will always be sending data regardless of being bound or not.

Reply:
Changed in commit 4